### ISSUE TITLE
np.random generates the same random numbers for each data batch

### ISSUE THREAD

**Issue by taoari** [pytorch/pytorch#5059]
For example, the following script 

```
from __future__ import print_function

import numpy as np
import random

# np.random.seed(0)

class Transform(object):
    def __init__(self):
        pass

    def __call__(self, item = None):
        return [np.random.randint(10000, 20000), random.randint(20000,30000)]

class RandomDataset(object):

    def __init__(self):
        pass

    def __getitem__(self, ind):
        item = [ind, np.random.randint(1, 10000), random.randint(10000, 20000), 0]
        tsfm =Transform()(item)
        return np.array(item + tsfm)
    def __len__(self):
        return 20


from torch.utils.data import DataLoader

ds = RandomDataset()
ds = DataLoader(ds, 10, shuffle=False, num_workers=4)

for batch in ds:
    print(batch)
```

gives 

```
#     0   2208  10983      0  15930  26264
#     1   2798  14403      0  17685  29545
#     2    528  16195      0  12927  28761
#     3   8541  13614      0  15240  24058
#     4   7144  14373      0  18374  28081
#     5   2329  17456      0  15192  26903
#     6    423  12168      0  18504  24193
#     7   9476  12027      0  19924  22325
#     8   3427  17570      0  12895  29773
#     9   6526  13327      0  15768  24566
#[torch.LongTensor of size 10x6]
#
#
#    10   2208  15203      0  15930  26024
#    11   2798  13264      0  17685  22011
#    12    528  11714      0  12927  24688
#    13   8541  11773      0  15240  29607
#    14   7144  14655      0  18374  24573
#    15   2329  12544      0  15192  27908
#    16    423  15892      0  18504  23111
#    17   9476  17389      0  19924  23799
#    18   3427  12458      0  12895  23201
#    19   6526  14935      0  15768  20789
#[torch.LongTensor of size 10x6]
```

The second and fifth columns are np.random generated random numbers in Dataset and Transform. However, they are always that the same for different batches. This will make random data augmentation actually "not random".

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @SsnL @VitalyFedyunin @ejguan @pbelevich

**Comment by SsnL** [comment]
This is because numpy doesn't properly handle RNG states when `fork` subprocesses. It's numpy's issue with multiprocessing tracked at https://github.com/numpy/numpy/issues/9248). But we do provide some workarounds, e.g. the `worker_init_fn` in DataLoader (see http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader). Or you can use other start methods like spawn.

**Comment by faustomilletari** [comment]
Hello,

what function should I pass to worker_init_fn to have bumpy generate true random numbers?

**Comment by SsnL** [comment]
@faustomilletari something like `lambda : np.random.seed(torch.initial_seed())`

**Comment by faustomilletari** [comment]
many thanks, 

I tried that but unfortunately it says that it is expecting numbers below 2^31 (or something along those lines). I even tried `lambda x: np.random.seed(torch.initial_seed() + x)` which would be more correct as it would seed each thread with a different seed. Same problem, of course.

Still, with both methods, even if they worked, as well as with `lambda x: np.random.seed(x)` the workers (in my case) produce exactly the same numbers again and again at every epoch. 

for me the only viable solution was `lambda _: np.random.seed()` which makes me lose control of the seed of those random sequences but at least ensures that same sequences "never" gets repeated.

**Comment by SsnL** [comment]
Well if it says that seed should be an int, you can just trim it as an int.
Adding to a number already too large isn’t the solution.

**Comment by ltrottier** [comment]
#### Solution (torch >= 0.4)

This is my solution:

1. First, reset the numpy seed at the beginning of each epoch:

```python
for epoch in range(2):
    print("epoch {}".format(epoch))
    np.random.seed() # reset seed
    for batch in ds:
        print(batch)
```

2. Use `worker_init_fn` to set the numpy seed of each worker:
```python
def worker_init_fn(worker_id):                                                          
    np.random.seed(np.random.get_state()[1][0] + worker_id)

ds = RandomDataset()
ds = DataLoader(ds, 10, shuffle=False, num_workers=4, worker_init_fn=worker_init_fn)
```

#### Explanations

After creating the workers, each worker has an independent seed that is initialized to the curent random seed + the id of the worker. Also, you need to reset the numpy random seed at the beginning of each epoch because all random seed modifications in `__getitem__` are local to each worker.

FYI, `np.random.get_state()[1][0]` allows you to get the seed. For example:
```python
In [1]: np.random.seed(1234)
   ...: np.random.get_state()[1][0]
   ...: 
Out[1]: 1234
``` 

#### Updated Code

```python
import numpy as np
import random

# np.random.seed(0)

class Transform(object):
    def __init__(self):
        pass

    def __call__(self, item = None):
        return [np.random.randint(10000, 20000), random.randint(20000,30000)]

class RandomDataset(object):

    def __init__(self):
        pass

    def __getitem__(self, ind):
        item = [ind, np.random.randint(1, 10000), random.randint(10000, 20000), 0]
        tsfm =Transform()(item)
        return np.array(item + tsfm)
    def __len__(self):
        return 20


from torch.utils.data import DataLoader

def worker_init_fn(worker_id):                                                          
    np.random.seed(np.random.get_state()[1][0] + worker_id)

ds = RandomDataset()
ds = DataLoader(ds, 10, shuffle=False, num_workers=4, worker_init_fn=worker_init_fn)

for epoch in range(2):
    print("epoch {}".format(epoch))
    np.random.seed()
    for batch in ds:
        print(batch)
```

**Comment by fcalvet** [comment]
Despite this being a NumPy issue, I feel the doc doesn't emphasize this strongly enough. Especially:
- the fact that the seed is the same at every epoch for every worker (one could easily assume workers are not epoch dependent and that the issue only affect randomness "in between workers" which could be a non issue for certain application)
- as the NumPy issue doesn't seem to happen on Windows, code could be performing fine on Windows and doing nonsense on Linux for example.

Furthermore, the example doesn't show how bad this can actually affect random generation (as for me it shows the issue for epoch reset and not in between workers)

```
import numpy as np
import torch
from torch.utils.data import Dataset, DataLoader

class FalseDataset(Dataset):
    def __init__(self, length):
        self.len = length

    def __len__(self):
        return self.len

    def __getitem__(self, ID):
        r = np.random.randint(1,10000)
        return [r, torch.initial_seed()]
    
false_instance = FalseDataset(8)
train_loader = DataLoader(false_instance, shuffle=False, num_workers=4)

def train_epoch(loader, epoch):
    for batch_idx, output in enumerate(loader):
        print(batch_idx, output)
    return "epoch"+str(epoch)+" ended"

for i in range(2): #simulating epochs
    print(train_epoch(train_loader, i))
```

which gives
```
0 [tensor([6441]), tensor([4899689041141450080])]
1 [tensor([6441]), tensor([4899689041141450081])]
2 [tensor([6441]), tensor([4899689041141450082])]
3 [tensor([6441]), tensor([4899689041141450083])]
4 [tensor([3807]), tensor([4899689041141450080])]
5 [tensor([3807]), tensor([4899689041141450081])]
6 [tensor([3807]), tensor([4899689041141450082])]
7 [tensor([3807]), tensor([4899689041141450083])]
epoch0 ended
0 [tensor([6441]), tensor([1246574829706935184])]
1 [tensor([6441]), tensor([1246574829706935185])]
2 [tensor([6441]), tensor([1246574829706935186])]
3 [tensor([6441]), tensor([1246574829706935187])]
4 [tensor([3807]), tensor([1246574829706935184])]
5 [tensor([3807]), tensor([1246574829706935185])]
6 [tensor([3807]), tensor([1246574829706935186])]
7 [tensor([3807]), tensor([1246574829706935187])]
epoch1 ended
```

**Comment by rainwang1995** [comment]
> #### Solution (torch >= 0.4)
> This is my solution:
> 
> 1. First, reset the numpy seed at the beginning of each epoch:
> 
> ```python
> for epoch in range(2):
>     print("epoch {}".format(epoch))
>     np.random.seed() # reset seed
>     for batch in ds:
>         print(batch)
> ```
> 
> 1. Use `worker_init_fn` to set the numpy seed of each worker:
> 
> ```python
> def worker_init_fn(worker_id):                                                          
>     np.random.seed(np.random.get_state()[1][0] + worker_id)
> 
> ds = RandomDataset()
> ds = DataLoader(ds, 10, shuffle=False, num_workers=4, worker_init_fn=worker_init_fn)
> ```
> 
> #### Explanations
> After creating the workers, each worker has an independent seed that is initialized to the curent random seed + the id of the worker. Also, you need to reset the numpy random seed at the beginning of each epoch because all random seed modifications in `__getitem__` are local to each worker.
> 
> FYI, `np.random.get_state()[1][0]` allows you to get the seed. For example:
> 
> ```python
> In [1]: np.random.seed(1234)
>    ...: np.random.get_state()[1][0]
>    ...: 
> Out[1]: 1234
> ```
> 
> #### Updated Code
> ```python
> import numpy as np
> import random
> 
> # np.random.seed(0)
> 
> class Transform(object):
>     def __init__(self):
>         pass
> 
>     def __call__(self, item = None):
>         return [np.random.randint(10000, 20000), random.randint(20000,30000)]
> 
> class RandomDataset(object):
> 
>     def __init__(self):
>         pass
> 
>     def __getitem__(self, ind):
>         item = [ind, np.random.randint(1, 10000), random.randint(10000, 20000), 0]
>         tsfm =Transform()(item)
>         return np.array(item + tsfm)
>     def __len__(self):
>         return 20
> 
> 
> from torch.utils.data import DataLoader
> 
> def worker_init_fn(worker_id):                                                          
>     np.random.seed(np.random.get_state()[1][0] + worker_id)
> 
> ds = RandomDataset()
> ds = DataLoader(ds, 10, shuffle=False, num_workers=4, worker_init_fn=worker_init_fn)
> 
> for epoch in range(2):
>     print("epoch {}".format(epoch))
>     np.random.seed()
>     for batch in ds:
>         print(batch)
> ```

Hi, I have tried the method you post. But for every epoch I get the same random number,
so it doesn't work for me.

**Comment by rainwang1995** [comment]
> many thanks,
> 
> I tried that but unfortunately it says that it is expecting numbers below 2^31 (or something along those lines). I even tried `lambda x: np.random.seed(torch.initial_seed() + x)` which would be more correct as it would seed each thread with a different seed. Same problem, of course.
> 
> Still, with both methods, even if they worked, as well as with `lambda x: np.random.seed(x)` the workers (in my case) produce exactly the same numbers again and again at every epoch.
> 
> for me the only viable solution was `lambda _: np.random.seed()` which makes me lose control of the seed of those random sequences but at least ensures that same sequences "never" gets repeated.

Thanks very much, lambda _: np.random.seed() works for me.

**Comment by faustomilletari** [comment]
it is incredible that after three years I return back to this post due to the same identical problem.

Most of the solution mentioned above work, but whenever there is random behaviour anywhere in the dataset, this problem is bound to happen (and it happens silently).

**Comment by SsnL** [comment]
> Still, with both methods, even if they worked, as well as with lambda x: np.random.seed(x) the workers (in my case) produce exactly the same numbers again and again at every epoch.

No, the default code should not give you the same `torch.initial_seed()` in worker for each epoch, unless you reseed the main process RNG at each epoch. The default code, at each epoch, fetches an integer from the main process RNG (which should be different for each epoch), and use that to seed the workers as `torch.manual_seed(seed + id)`. Then, inside the `worker_init_fn`, `torch.initial_seed()` would give you this new seed, which is different for different epochs.

**Comment by SsnL** [comment]
@rainwang1995 see
```py
In [3]: import torch
   ...:
   ...: class Dataset(object):
   ...:     def __len__(self): return 20
   ...:     def __getitem__(self, k): return k
   ...:
   ...: def wif(id):
   ...:     print(id, torch.initial_seed(), flush=True)
   ...:
   ...: loader = torch.utils.data.DataLoader(Dataset(), num_workers=5, batch_size=2, worker
   ...: _init_fn=wif)
   ...:
   ...: for e in range(5):
   ...:     print(f"\nepoch {e}")
   ...:     for _ in loader:
   ...:         pass
   ...:

epoch 0
0 2058943141876568872
1 2058943141876568873
2 2058943141876568874
3 2058943141876568875
4 2058943141876568876

epoch 1
0 8513519487925896339
1 8513519487925896340
2 8513519487925896341
3 8513519487925896342
4 8513519487925896343

epoch 2
0 4095240921208205829
1 4095240921208205830
2 4095240921208205831
3 4095240921208205832
4 4095240921208205833

epoch 3
0 8839229708335947710
1 8839229708335947711
2 8839229708335947712
3 8839229708335947713
4 8839229708335947714

epoch 4
0 45077338359890785
1 45077338359890786
2 45077338359890787
3 45077338359890788
4 45077338359890789
```

**Comment by konioyxgq** [comment]
@SsnL Why the **spawn** method works?

**Comment by SsnL** [comment]
@konioyxgq Because fork conceptually share some process state with the parent process, which include RNG state. But spawn starts a new process, which does not share things with the parent.

**Comment by konioyxgq** [comment]
Thank you very much for your reply. @SsnL

**Comment by tanelp** [comment]
I want to emphasize that this is a common problem.

I downloaded and analysed over a hundred thousand repositories from GitHub that import PyTorch. I kept projects that use NumPy’s random number generator with multi-process data loading. Out of these, over 95% of the repositories are plagued by this problem. It’s inside PyTorch’s official tutorial, OpenAI’s code, NVIDIA’s projects, etc.

You can read more from [here](https://tanelp.github.io/posts/a-bug-that-plagues-thousands-of-open-source-ml-projects/).

**Comment by rkern** [comment]
Hi there, original `numpy.random` author here. I'm happy to talk with anyone about how to incorporate numpy's [recommended patterns for parallel PRNG use](https://numpy.org/doc/stable/reference/random/parallel.html) into PyTorch's documentation and/or code.

Please accept my commiserations for any time spent debugging issues related to this.

**Comment by lucmos** [comment]
Hello there,
My current workaround adapted from @SsnL [answer](https://github.com/pytorch/pytorch/issues/5059#issuecomment-624788105) is the following:

```python
import numpy as np
import torch

# torch.manual_seed(0)


class Dataset(object):
    def __len__(self):
        return 10

    def __getitem__(self, k):
        return f"seed({np.random.get_state()[1][0]})"


def wif(id):
    np.random.seed((id + torch.initial_seed()) % np.iinfo(np.int32).max)


loader = torch.utils.data.DataLoader(
    Dataset(), num_workers=5, batch_size=2, worker_init_fn=wif
)
for e in range(2):
    print(f"\nepoch {e}")
    for i, batch in enumerate(loader):
        print(f"batch: {i}, {batch}")
```

```
epoch 0
batch: 0, ['seed(2755900635)', 'seed(2755900635)']
batch: 1, ['seed(2755900637)', 'seed(2755900637)']
batch: 2, ['seed(2755900639)', 'seed(2755900639)']
batch: 3, ['seed(2755900641)', 'seed(2755900641)']
batch: 4, ['seed(2755900643)', 'seed(2755900643)']

epoch 1
batch: 0, ['seed(255357750)', 'seed(255357750)']
batch: 1, ['seed(255357752)', 'seed(255357752)']
batch: 2, ['seed(255357754)', 'seed(255357754)']
batch: 3, ['seed(255357756)', 'seed(255357756)']
batch: 4, ['seed(255357758)', 'seed(255357758)']
```

It seems to work and to be deterministic between runs if a global `torch.manual_seed(0)` is set. However, I am not making use of the recommended numpy patterns for parallel PRNG use. 

Am I hitting any hard-to-spot bugs? Which is the suggested alternative?

**Comment by nik-sm** [comment]
In case it is useful, you can also find this problem described in the docs here: https://pytorch.org/docs/stable/data.html#randomness-in-multi-process-data-loading, and the solution mentioned here: https://pytorch.org/docs/stable/notes/randomness.html#dataloader

**Comment by rkern** [comment]
_Ideally_, I recommend that people create and pass around explicit `RandomState/Generator` instances. However, given the current API of `DataLoader` (and wanting a quick fix for the code that's in the wild), let's instead start under the constraint that you want to reseed the implicit global `RandomState` underneath the `np.random` convenience functions. This will make the solutions uglier than they ought to be, but let's see what we can do.

So the most minimal option is to make the call to `seed()` safer, though still within the constraints of the old `MT19937` legacy seeding algorithm (which is still serviceable). The thing to note is that it doesn't require everything to be jammed down into a single 32-bit integer. It can take an arbitrary list of `uint32` integers. `torch.initial_seed()` uses a `uint64`-sized integer, so we need to break it up, and concatenate the worker ID rather than add it. However, once we're in the `worker_init_fn` in the child process, the `torch.initial_seed()` has already been reset to a number derived from the parent's state (drawing a `base_seed` integer and then adding the worker ID). So we just need to handle the `uint64`-sized integer, not the worker ID.

BTW, when printing out the results of `get_state()` to check for consistency, be sure to look at the second element of the state array, `get_state()[1][1]`. The first element is actually irrelevant; the consistent variations that you were seeing are just an accident of the implementation details.

```python
def wif(id):
    uint64_seed = torch.initial_seed()
    np.random.seed([uint64_seed >> 32, uint64_seed & 0xffff_ffff])
```

The new `SeedSequence` provides a state-derivation algorithm that has some better properties, though it's more beneficial for the spawning API that we're not really able to use here. But it does handle lists of arbitrary-sized integers with less fuss, so let's do something like the above with `SeedSequence` instead.

```python
def wif(id):
    uint64_seed = torch.initial_seed()
    ss = np.random.SeedSequence([uint64_seed])
    # More than 128 bits (4 32-bit words) would be overkill.
    np.random.seed(ss.generate_state(4))
```

If we want to get fancy, we can undo the seed derivation that `DataLoader` did and recover the `base_seed` so we can use all of the bits that are available to us. Assuming that the seed derivation logic doesn't change out from under us, this is the probably the highest-quality version, given the constraints.

```python
def wif(id):
    process_seed = torch.initial_seed()
    # Back out the base_seed so we can use all the bits.
    base_seed = process_seed - id
    ss = np.random.SeedSequence([id, base_seed])
    # More than 128 bits (4 32-bit words) would be overkill.
    np.random.seed(ss.generate_state(4))
```

I think those are pretty reasonable quick fixes for people using `np.random` inside of their `Dataset` objects. I do think that `DataLoader` might consider using `SeedSequence` (either directly or vendorized) for deriving the seeds that it computes and passes to its workers for `torch.random` and the stdlib `random`. There are two potential problems that I see with the way that it's currently done (generating a new `base_seed` from the existing PRNG, and then deriving worker seeds my adding the worker IDs to that base). Judging from the documentation warning against using small seed integers with an imbalance of 1 and 0 bits, if that's accurate, then you probably want the protection of a good seed derivation algorithm like `SeedSequence` to condition the inputs anyways. But those would break reproducibility across versions of PyTorch, so I'll only dive into them if there's appetite for such a change.

**Comment by KyotoSunshine** [comment]
@rkern Since torch.initial_seed() is only 64 bits, why do we need 128 bits to set np.random.seed? Is 64 bits not enough? I.e. two 32 bit words.

**Comment by rgommers** [comment]
> Judging from the documentation warning against using small seed integers with an imbalance of 1 and 0 bits, if that's accurate, then you probably want the protection of a good seed derivation algorithm like `SeedSequence` to condition the inputs anyways. But those would break reproducibility across versions of PyTorch, so I'll only dive into them if there's appetite for such a change.

Such reproducibility isn't guaranteed anyway, it's the first sentence of https://pytorch.org/docs/master/notes/randomness.html#reproducibility. So this may be worth considering - @mruberry can you comment on this?

**Comment by mruberry** [comment]
> > Judging from the documentation warning against using small seed integers with an imbalance of 1 and 0 bits, if that's accurate, then you probably want the protection of a good seed derivation algorithm like `SeedSequence` to condition the inputs anyways. But those would break reproducibility across versions of PyTorch, so I'll only dive into them if there's appetite for such a change.
> 
> Such reproducibility isn't guaranteed anyway, it's the first sentence of https://pytorch.org/docs/master/notes/randomness.html#reproducibility. So this may be worth considering - @mruberry can you comment on this?

It's true that we don't guarantee reproducibility across PyTorch versions, but @VitalyFedyunin is looking at updating our dataloader API now, so he probably has a better sense for issues dataloading is sensitive to.

**Comment by VitalyFedyunin** [comment]
@rkern thank you for detailed instruction how to reseed NumPy.

It is not hard to update PyTorch DataLoader to perform such reseeding for every individual worker. But we had two reasons not to do it: 1) naive implementation is going to be backward compatibility breaking change and might impact people who write tests 2) even if we do it for NumPy every other 3rd party library would be still impacted. 

However, because of the problem scale, we are going to make exception and implement such reseeding inside of our current DataLoader. We will also add global context (tied to #53532 feature) to disable such functionality in case of users need.

**Comment by ejguan** [comment]
### TL;DR
PR #56488 will fix this issue. The DataLoader will perform re-seeding for NumPy at each worker by default. And, it won't impact users who have already used `worker_init_fn` to do their customized re-seeding.

### Problem Statement
And, I want to summarize several facts for this problem, just in case that users who use other third-party library for random operation within each worker (Dataset) encounter the same issue for random seeding.
There are two main concerns for random seeding of 3rd party library in DataLoader without explicitly specified seed by `worker_init_fn`.
- Same seed across different workers -> same random result at each iteration
- Same initial seed for each epoch

These two problems happen when `fork` is set as the start method for multiprocessing, because the NumPy seed of each worker would inherit same NumPy seed from the main process. And, based on [the document of Python multiprocessing]((https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods), **`fork` is the default method on Unix platform**. So, users need to take more care of it on Unix, such as to specify the start method for multiprocessing to `spawn` or to use `worker_init_fn` to specify seed for the 3rd party library (recommended). 
But, it doesn't mean users could ignore the seeding on other platforms. In order to keep the reproducibility, users still need to care about their seeding strategy.

### Reproducibility:
There are two main seeds within DataLoader to control the reproducibility.
- Worker seed
Torch, NumPy and random will have a seed by default, which can be reproduced with or without `worker_init_fn`
- Base seed
Users usually ignore this seed. But, it's important for reproducibility as all the worker seeds are derived from this base seed. So, there are two ways to control the base seed.
  - Set `torch.manual_seed(base_seed)` before DataLoader
  - Explicitly set RNG for DataLoader
```py
generator = torch. Generator()
g.manual_seed(base_seed)
DataLoader(dataset, ..., generator=generator)
```

### Request for Feedback
We really appreciate your feedback to improve the DataLoader.
And, I hope to gather some feedback on the reproducibility for data loading process. During your training process, do you or how often do you need to reproduce the data flow?

**Comment by AnkushMalaker** [comment]
@ezyang I'm using `imgaug` with pytorch lightning. The augmentations for each epoch are the same. Is it related to this issue? 
Sorry for tagging you, I'm having a hard time finding where to look for a fix for this issue/identifying the cause.

**Comment by pGit1** [comment]
Hi @AnkushMalaker , don't know if you still need an answer but please look here: https://github.com/aleju/imgaug/issues/406#issuecomment-670814426

**Comment by RylanSchaeffer** [comment]
@rkern  when I try your solutions, they work for 1 dataloader + dataset, but if I create different datasets and different dataloaders, that `wif` returns the same sequence of numpy sampled data in each dataset. I tried adding a `split_seed` argument so I can have a different, reproducible sequence of data for my three splits (train, val, test) like so:

```
    def worker_init_fn(id, split_seed: int):
        # Recommended by NumPy Rng Author: https://github.com/pytorch/pytorch/issues/5059#issuecomment-817392562
        # Another good resource: https://tanelp.github.io/posts/a-bug-that-plagues-thousands-of-open-source-ml-projects/
        process_seed = torch.initial_seed()
        # Back out the base_seed so we can use all the bits.
        base_seed = process_seed - id
        # TODO: split_seed seems to have no impact.
        ss = np.random.SeedSequence(
            [id, base_seed, split_seed]
        )  # Rylan added split seed.
        # More than 128 bits (4 32-bit words) would be overkill.
        np.random.seed(ss.generate_state(4))
```

But this appears to have no impact. What would you recommend?

**Comment by rkern** [comment]
It works fine for me in `torch==2.0.1` even without the `split_seed` because a different `base_seed` will be pulled from the PyTorch global PRNG for each `DataLoader` in any case. But even if I force the same PyTorch PRNG state with `torch.manual_seed(seed)` before the loop over each `DataLoader`, then the `split_seed` works fine.

https://gist.github.com/rkern/4cd064617a16b93554ec91704c3f3f14

If you can post your own complete minimal example as a gist, we can continue to discuss over there.

**Comment by RylanSchaeffer** [comment]
I investigated and I think the bug is actually from another library (Pytorch Lightning). All good - thank you for investigating!

**Comment by adosar** [comment]
> No, the default code should not give you the same torch.initial_seed() in worker for each epoch, unless you reseed the main process RNG at each epoch. The default code, at each epoch, fetches an integer from the main process RNG ...

@ssnl So datasets that use `torch` for randomness should be OK without a `worker_init_fn`? Do you know where in the source code this fetching of integer happens?
